In this week, we will use basketball data downloaded from NBA.com to demonstrate how to import data into Python, how to clean up data before conducting any data analyses, as well how to describe and summarize data.

In [1]:
import pandas as pd
import numpy as np

In [2]:
nba_teams=pd.read_csv("Data/nba_teams.csv")

In [3]:
display(nba_teams)

Unnamed: 0 ABBREVIATION           CITY               FULL_NAME  \
0            0          ATL        Atlanta           Atlanta Hawks   
1            1          BOS         Boston          Boston Celtics   
2            2          CLE      Cleveland     Cleveland Cavaliers   
3            3          NOP    New Orleans    New Orleans Pelicans   
4            4          CHI        Chicago           Chicago Bulls   
5            5          DAL         Dallas        Dallas Mavericks   
6            6          DEN         Denver          Denver Nuggets   
7            7          GSW   Golden State   Golden State Warriors   
8            8          HOU        Houston         Houston Rockets   
9            9          LAC    Los Angeles    Los Angeles Clippers   
10          10          LAL    Los Angeles      Los Angeles Lakers   
11          11          MIA          Miami              Miami Heat   
12          12          MIL      Milwaukee         Milwaukee Bucks   
13          13          MIN      Minnesota  Minnesota Timberwolves   
14          14          BKN       Brooklyn           Brooklyn Nets   
15          15          NYK       New York         New York Knicks   
16          16          ORL        Orlando           Orlando Magic   
17          17          IND        Indiana          Indiana Pacers   
18          18          PHI   Philadelphia      Philadelphia 76ers   
19          19          PHX        Phoenix            Phoenix Suns   
20          20          POR       Portland  Portland Trail Blazers   
21          21          SAC     Sacramento        Sacramento Kings   
22          22          SAS    San Antonio       San Antonio Spurs   
23          23          OKC  Oklahoma City   Oklahoma City Thunder   
24          24          TOR        Toronto         Toronto Raptors   
25          25          UTA           Utah               Utah Jazz   
26          26          MEM        Memphis       Memphis Grizzlies   
27          27          WAS     Washington      Washington Wizards   
28          28          DET        Detroit         Detroit Pistons   
29          29          CHA      Charlotte       Charlotte Hornets   

            ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz                  Utah          1974

This dataset provides some basic information of the NBA teams.

For a dataset, each row represents an observation, i.e., a team in this dataset and each column represents a variable which contains information of a characteristics of the observation. A variable can take different values in different situations. The number of observation in a dataset represents the size of our sample and the number of variables represents the richness of information in our dataset.

We can use the “shape” function in Python to see how many variables and observations in our dataset.

In [4]:
nba_teams.shape

(30, 8)

In [5]:
nba_teams.rename(columns={'Unnamed: 0':'TEAM_NUMBER'}, inplace=True)

nba_teams.rename(columns={'ID':'TEAM_ID'}, inplace=True)
display(nba_teams)

TEAM_NUMBER ABBREVIATION           CITY               FULL_NAME  \
0             0          ATL        Atlanta           Atlanta Hawks   
1             1          BOS         Boston          Boston Celtics   
2             2          CLE      Cleveland     Cleveland Cavaliers   
3             3          NOP    New Orleans    New Orleans Pelicans   
4             4          CHI        Chicago           Chicago Bulls   
5             5          DAL         Dallas        Dallas Mavericks   
6             6          DEN         Denver          Denver Nuggets   
7             7          GSW   Golden State   Golden State Warriors   
8             8          HOU        Houston         Houston Rockets   
9             9          LAC    Los Angeles    Los Angeles Clippers   
10           10          LAL    Los Angeles      Los Angeles Lakers   
11           11          MIA          Miami              Miami Heat   
12           12          MIL      Milwaukee         Milwaukee Bucks   
13           13          MIN      Minnesota  Minnesota Timberwolves   
14           14          BKN       Brooklyn           Brooklyn Nets   
15           15          NYK       New York         New York Knicks   
16           16          ORL        Orlando           Orlando Magic   
17           17          IND        Indiana          Indiana Pacers   
18           18          PHI   Philadelphia      Philadelphia 76ers   
19           19          PHX        Phoenix            Phoenix Suns   
20           20          POR       Portland  Portland Trail Blazers   
21           21          SAC     Sacramento        Sacramento Kings   
22           22          SAS    San Antonio       San Antonio Spurs   
23           23          OKC  Oklahoma City   Oklahoma City Thunder   
24           24          TOR        Toronto         Toronto Raptors   
25           25          UTA           Utah               Utah Jazz   
26           26          MEM        Memphis       Memphis Grizzlies   
27           27          WAS     Washington      Washington Wizards   
28           28          DET        Detroit         Detroit Pistons   
29           29          CHA      Charlotte       Charlotte Hornets   

       TEAM_ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz     

Self Test - 1
Rename "FULL_NAME" to "TEAM_NAME"

In [6]:
nba_teams.rename(columns=dict(FULL_NAME="TEAM_NAME"), inplace=True)
display(nba_teams)

TEAM_NUMBER ABBREVIATION           CITY               TEAM_NAME  \
0             0          ATL        Atlanta           Atlanta Hawks   
1             1          BOS         Boston          Boston Celtics   
2             2          CLE      Cleveland     Cleveland Cavaliers   
3             3          NOP    New Orleans    New Orleans Pelicans   
4             4          CHI        Chicago           Chicago Bulls   
5             5          DAL         Dallas        Dallas Mavericks   
6             6          DEN         Denver          Denver Nuggets   
7             7          GSW   Golden State   Golden State Warriors   
8             8          HOU        Houston         Houston Rockets   
9             9          LAC    Los Angeles    Los Angeles Clippers   
10           10          LAL    Los Angeles      Los Angeles Lakers   
11           11          MIA          Miami              Miami Heat   
12           12          MIL      Milwaukee         Milwaukee Bucks   
13           13          MIN      Minnesota  Minnesota Timberwolves   
14           14          BKN       Brooklyn           Brooklyn Nets   
15           15          NYK       New York         New York Knicks   
16           16          ORL        Orlando           Orlando Magic   
17           17          IND        Indiana          Indiana Pacers   
18           18          PHI   Philadelphia      Philadelphia 76ers   
19           19          PHX        Phoenix            Phoenix Suns   
20           20          POR       Portland  Portland Trail Blazers   
21           21          SAC     Sacramento        Sacramento Kings   
22           22          SAS    San Antonio       San Antonio Spurs   
23           23          OKC  Oklahoma City   Oklahoma City Thunder   
24           24          TOR        Toronto         Toronto Raptors   
25           25          UTA           Utah               Utah Jazz   
26           26          MEM        Memphis       Memphis Grizzlies   
27           27          WAS     Washington      Washington Wizards   
28           28          DET        Detroit         Detroit Pistons   
29           29          CHA      Charlotte       Charlotte Hornets   

       TEAM_ID       NICKNAME                 STATE  YEAR_FOUNDED  
0   1610612737          Hawks               Atlanta          1949  
1   1610612738        Celtics         Massachusetts          1946  
2   1610612739      Cavaliers                  Ohio          1970  
3   1610612740       Pelicans             Louisiana          2002  
4   1610612741          Bulls              Illinois          1966  
5   1610612742      Mavericks                 Texas          1980  
6   1610612743        Nuggets              Colorado          1976  
7   1610612744       Warriors            California          1946  
8   1610612745        Rockets                 Texas          1967  
9   1610612746       Clippers            California          1970  
10  1610612747         Lakers            California          1948  
11  1610612748           Heat               Florida          1988  
12  1610612749          Bucks             Wisconsin          1968  
13  1610612750   Timberwolves             Minnesota          1989  
14  1610612751           Nets              New York          1976  
15  1610612752         Knicks              New York          1946  
16  1610612753          Magic               Florida          1989  
17  1610612754         Pacers               Indiana          1976  
18  1610612755          76ers          Pennsylvania          1949  
19  1610612756           Suns               Arizona          1968  
20  1610612757  Trail Blazers                Oregon          1970  
21  1610612758          Kings            California          1948  
22  1610612759          Spurs                 Texas          1976  
23  1610612760        Thunder              Oklahoma          1967  
24  1610612761        Raptors               Ontario          1995  
25  1610612762           Jazz     

### Dropping Variables (columns)

### To drop a variable, i.e., to delete a column, we can use the “drop” command.¶
- We need to provide the name of the variable;
- We also need to use the argument “axis=1” which tells Python that we are dropping a column, not a row.

The variable "TEAM_NUMBER" has little meaning, let's drop 

In [7]:
nba_teams.drop(['TEAM_NUMBER'], axis=1, inplace=True)
display(nba_teams)

ABBREVIATION           CITY               TEAM_NAME     TEAM_ID  \
0           ATL        Atlanta           Atlanta Hawks  1610612737   
1           BOS         Boston          Boston Celtics  1610612738   
2           CLE      Cleveland     Cleveland Cavaliers  1610612739   
3           NOP    New Orleans    New Orleans Pelicans  1610612740   
4           CHI        Chicago           Chicago Bulls  1610612741   
5           DAL         Dallas        Dallas Mavericks  1610612742   
6           DEN         Denver          Denver Nuggets  1610612743   
7           GSW   Golden State   Golden State Warriors  1610612744   
8           HOU        Houston         Houston Rockets  1610612745   
9           LAC    Los Angeles    Los Angeles Clippers  1610612746   
10          LAL    Los Angeles      Los Angeles Lakers  1610612747   
11          MIA          Miami              Miami Heat  1610612748   
12          MIL      Milwaukee         Milwaukee Bucks  1610612749   
13          MIN      Minnesota  Minnesota Timberwolves  1610612750   
14          BKN       Brooklyn           Brooklyn Nets  1610612751   
15          NYK       New York         New York Knicks  1610612752   
16          ORL        Orlando           Orlando Magic  1610612753   
17          IND        Indiana          Indiana Pacers  1610612754   
18          PHI   Philadelphia      Philadelphia 76ers  1610612755   
19          PHX        Phoenix            Phoenix Suns  1610612756   
20          POR       Portland  Portland Trail Blazers  1610612757   
21          SAC     Sacramento        Sacramento Kings  1610612758   
22          SAS    San Antonio       San Antonio Spurs  1610612759   
23          OKC  Oklahoma City   Oklahoma City Thunder  1610612760   
24          TOR        Toronto         Toronto Raptors  1610612761   
25          UTA           Utah               Utah Jazz  1610612762   
26          MEM        Memphis       Memphis Grizzlies  1610612763   
27          WAS     Washington      Washington Wizards  1610612764   
28          DET        Detroit         Detroit Pistons  1610612765   
29          CHA      Charlotte       Charlotte Hornets  1610612766   

         NICKNAME                 STATE  YEAR_FOUNDED  
0           Hawks               Atlanta          1949  
1         Celtics         Massachusetts          1946  
2       Cavaliers                  Ohio          1970  
3        Pelicans             Louisiana          2002  
4           Bulls              Illinois          1966  
5       Mavericks                 Texas          1980  
6         Nuggets              Colorado          1976  
7        Warriors            California          1946  
8         Rockets                 Texas          1967  
9        Clippers            California          1970  
10         Lakers            California          1948  
11           Heat               Florida          1988  
12          Bucks             Wisconsin          1968  
13   Timberwolves             Minnesota          1989  
14           Nets              New York          1976  
15         Knicks              New York          1946  
16          Magic               Florida          1989  
17         Pacers               Indiana          1976  
18          76ers          Pennsylvania          1949  
19           Suns               Arizona          1968  
20  Trail Blazers                Oregon          1970  
21          Kings            California          1948  
22          Spurs                 Texas          1976  
23        Thunder              Oklahoma          1967  
24        Raptors               Ontario          1995  
25           Jazz                  Utah          1974  
26      Grizzlies             Tennessee          1995  
27        Wizards  District of Columbia          1961  
28        Pistons              Michigan          1948  
29        Hornets        North Carolina          1988

#### Next we will work on game level data.

Import the game level dataset from our data repository. 
- We can display just first five rows of the dataset using the "head" command.

In [28]:
games=pd.read_csv("Data/basketball_games.csv")
games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
0      22019  1611661322               WAS  Washington Mystics  1021900079   
1      22019  1611661319               LVA      Las Vegas Aces  1021900079   
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   

  GAME_DATE      MATCHUP   WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  \
0  8/5/2019    WAS @ LVA  NaN  101   51  ...   0.750     3    10   13   14   
1  8/5/2019  LVA vs. WAS  NaN  100   36  ...   0.800     5    10   15   12   
2  8/1/2019  LAS vs. LVA    W  200   76  ...   1.000     5    28   33   19   
3  8/1/2019  CON vs. PHO    W  200   68  ...   0.786    12    28   40   18   
4  8/1/2019    PHO @ CON    L  199   62  ...   0.778     6    23   29   15   

   STL  BLK  TOV  PF  PLUS_MINUS  
0    2    4    5   7        12.2  
1    2    0    7   8       -11.8  
2    7    7   12  18         8.0  
3   13    0   16  15         6.0  
4    8    9   18  11        -6.0  

[5 rows x 28 columns]

_Upon importing the game data, we notice that the first five games are not NBA games, instead, they are WNBA games. Indeed, this dataset contains NBA games, WNBA games, NBA 2K (simulation video) games._ 

### Dropping observations (rows)

#### To drop an observation, we can use the index number on the left to specify the row we want to drop.
- The argument axis=0 specifies that we want to drop a row instead of a column.


In [29]:
games.drop([0], axis=0, inplace=True)
games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
1      22019  1611661319               LVA      Las Vegas Aces  1021900079   
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   
5      22019  1611661319               LVA      Las Vegas Aces  1021900124   

  GAME_DATE      MATCHUP   WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  \
1  8/5/2019  LVA vs. WAS  NaN  100   36  ...   0.800     5    10   15   12   
2  8/1/2019  LAS vs. LVA    W  200   76  ...   1.000     5    28   33   19   
3  8/1/2019  CON vs. PHO    W  200   68  ...   0.786    12    28   40   18   
4  8/1/2019    PHO @ CON    L  199   62  ...   0.778     6    23   29   15   
5  8/1/2019    LVA @ LAS    L  200   68  ...   0.765    10    29   39   17   

   STL  BLK  TOV  PF  PLUS_MINUS  
1    2    0    7   8       -11.8  
2    7    7   12  18         8.0  
3   13    0   16  15         6.0  
4    8    9   18  11        -6.0  
5    7    4   14  13        -8.0  

[5 rows x 28 columns]

#### More often, we will drop observations based on certain conditions. 

For example, Las Vegas Aces is a women’s basketball team. If we are only going to focus on men’s basketball games, we will drop all the games played by Las Vegas Aces. In this case, we don’t have to use the “drop” function. We can specify our TEAM_NAME variables to be not equal to “Las Vegas Aces.”

In [30]:
games=games[games.TEAM_NAME !="Las Vegas Aces"]
games.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
4      22019  1611661317               PHO     Phoenix Mercury  1021900122   
6      22019  1611661313               NYL    New York Liberty  1021900123   
7      22019  1611661321               DAL        Dallas Wings  1021900123   

  GAME_DATE      MATCHUP WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  STL  \
2  8/1/2019  LAS vs. LVA  W  200   76  ...   1.000     5    28   33   19    7   
3  8/1/2019  CON vs. PHO  W  200   68  ...   0.786    12    28   40   18   13   
4  8/1/2019    PHO @ CON  L  199   62  ...   0.778     6    23   29   15    8   
6  8/1/2019    NYL @ DAL  L  200   64  ...   0.800     6    21   27   15   10   
7  8/1/2019  DAL vs. NYL  W  200   87  ...   0.833     9    24   33   24   13   

   BLK  TOV  PF  PLUS_MINUS  
2    7   12  18         8.0  
3    0   16  15         6.0  
4    9   18  11        -6.0  
6    2   21  26       -23.0  
7    6   18  21        23.0  

[5 rows x 28 columns]

## Self Test - 2
- Drop all the Phoenix Mercury games

In [32]:
mercury_games = games[games.TEAM_NAME=="Phoenix Mercury"]
len(mercury_games)

271

In [33]:
print(games.shape)
games = games[games.TEAM_NAME != "Phoenix Mercury"]
print(games.shape)
games.head()

(29941, 28)
(29670, 28)


SEASON_ID     TEAM_ID TEAM_ABBREVIATION           TEAM_NAME     GAME_ID  \
2      22019  1611661320               LAS  Los Angeles Sparks  1021900124   
3      22019  1611661323               CON     Connecticut Sun  1021900122   
6      22019  1611661313               NYL    New York Liberty  1021900123   
7      22019  1611661321               DAL        Dallas Wings  1021900123   
8      22019  1611661325               IND       Indiana Fever  1021900121   

   GAME_DATE      MATCHUP WL  MIN  PTS  ...  FT_PCT  OREB  DREB  REB  AST  \
2   8/1/2019  LAS vs. LVA  W  200   76  ...   1.000     5    28   33   19   
3   8/1/2019  CON vs. PHO  W  200   68  ...   0.786    12    28   40   18   
6   8/1/2019    NYL @ DAL  L  200   64  ...   0.800     6    21   27   15   
7   8/1/2019  DAL vs. NYL  W  200   87  ...   0.833     9    24   33   24   
8  7/31/2019  IND vs. ATL  W  200   61  ...   0.731     7    36   43   14   

   STL  BLK  TOV  PF  PLUS_MINUS  
2    7    7   12  18         8.0  
3   13    0   16  15         6.0  
6   10    2   21  26       -23.0  
7   13    6   18  21        23.0  
8   11    4   15  17         2.0  

[5 rows x 28 columns]

## Merging Dataframes

#### We will only focus on NBA games. We could merge the NBA_Teams and Games datasets to filter out NBA games.

#### Teams are identified by the TEAM_ID. So, let’s merge the datasets by TEAM_ID. Since the variable “TEAM_NAME” is also present in both datasets, we could also include this variable as a criteria to merge the datasets so that in our new dataset, there is no duplicate variables.

In [47]:
nba_data = pd.merge(nba_teams, games, on=["TEAM_ID", "TEAM_NAME"])
not_nba = games[~games.TEAM_NAME.isin(nba_teams.TEAM_NAME)]
print(f"Shape of not nba data: {not_nba.shape}")
nba_data

Shape of not nba data: (11249, 28)


ABBREVIATION       CITY          TEAM_NAME     TEAM_ID NICKNAME  \
0              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
1              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
2              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
3              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
4              ATL    Atlanta      Atlanta Hawks  1610612737    Hawks   
...            ...        ...                ...         ...      ...   
18416          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18417          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18418          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18419          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   
18420          CHA  Charlotte  Charlotte Hornets  1610612766  Hornets   

                STATE  YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  \
0             Atlanta          1949      22019               ATL  1521900072   
1             Atlanta          1949      22019               ATL  1521900060   
2             Atlanta          1949      22019               ATL  1521900042   
3             Atlanta          1949      22019               ATL  1521900023   
4             Atlanta          1949      22019               ATL  1521900023   
...               ...           ...        ...               ...         ...   
18416  North Carolina          1988      22014               CHA  1521400046   
18417  North Carolina          1988      22014               CHA  1521400038   
18418  North Carolina          1988      22014               CHA  1521400023   
18419  North Carolina          1988      22014               CHA  1521400017   
18420  North Carolina          1988      22014               CHA  1521400004   

       ... FT_PCT OREB DREB  REB  AST  STL  BLK  TOV  PF  PLUS_MINUS  
0      ...  0.850   13   23   36   14   15    3   12  24         8.0  
1      ...  0.700    9   28   37   19   10    8   22  25        -5.0  
2      ...  0.708    7   27   34   17    5    5   18  21        18.2  
3      ...  0.500    1    2    3    1    0    1    4   2         0.0  
4      ...  0.625    9   27   36    7    7   10   18  28       -24.0  
...    ...    ...  ...  ...  ...  ...  ...  ...  ...  ..         ...  
18416  ...  0.762   12   26   38   19    7    2   11  22        29.0  
18417  ...  0.667    7   33   40   10    7    2   21  31         1.0  
18418  ...  0.618    4   19   23   10    5    5   16  19       -23.0  
18419  ...  0.615   19   28   47    9    5    1   17  39        -7.0  
18420  ...  0.789   14   20   34    4    8    8   14  20       -12.0  

[18421 rows x 33 columns]

### Understanding and cleaning the merged dataset

_As you can tell, the merged dataset has a lot more variables and Python cannot fit all of them in the screen._ 

#### We can obtain the list of variables using the “columns” command. This provides us a full list of variables in our dataset.

In [49]:
nba_data.columns

Index(['ABBREVIATION', 'CITY', 'TEAM_NAME', 'TEAM_ID', 'NICKNAME', 'STATE',
       'YEAR_FOUNDED', 'SEASON_ID', 'TEAM_ABBREVIATION', 'GAME_ID',
       'GAME_DATE', 'MATCHUP', 'WL', 'MIN', 'PTS', 'FGM', 'FGA', 'FG_PCT',
       'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA', 'FT_PCT', 'OREB', 'DREB',
       'REB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PLUS_MINUS'],
      dtype='object')

#### Data Cleaning

The variable “ABBREVIATION” AND “TEAM_ABBREVIATION” carry the same information and it is not necessary to keep both of them.

- Delete "ABBREVIATION"
- using errors="ignore" option makes code re-runnable as already missing column won't raise KeyError

In [59]:
nba_data.drop(["ABBREVIATION"], axis=1, inplace=True, errors="ignore")
nba_data

CITY          TEAM_NAME     TEAM_ID NICKNAME           STATE  \
0        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
1        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
2        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
3        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
4        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
...          ...                ...         ...      ...             ...   
18416  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18417  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18418  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18419  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18420  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ...  \
0              1949      22019               ATL  1521900072  7/12/2019  ...   
1              1949      22019               ATL  1521900060  7/11/2019  ...   
2              1949      22019               ATL  1521900042   7/9/2019  ...   
3              1949      22019               ATL  1521900023   7/7/2019  ...   
4              1949      22019               ATL  1521900023   7/7/2019  ...   
...             ...        ...               ...         ...        ...  ...   
18416          1988      22014               CHA  1521400046  7/17/2014  ...   
18417          1988      22014               CHA  1521400038  7/16/2014  ...   
18418          1988      22014               CHA  1521400023  7/14/2014  ...   
18419          1988      22014               CHA  1521400017  7/13/2014  ...   
18420          1988      22014               CHA  1521400004  7/11/2014  ...   

      FT_PCT OREB  DREB  REB  AST  STL  BLK  TOV  PF  PLUS_MINUS  
0      0.850   13    23   36   14   15    3   12  24         8.0  
1      0.700    9    28   37   19   10    8   22  25        -5.0  
2      0.708    7    27   34   17    5    5   18  21        18.2  
3      0.500    1     2    3    1    0    1    4   2         0.0  
4      0.625    9    27   36    7    7   10   18  28       -24.0  
...      ...  ...   ...  ...  ...  ...  ...  ...  ..         ...  
18416  0.762   12    26   38   19    7    2   11  22        29.0  
18417  0.667    7    33   40   10    7    2   21  31         1.0  
18418  0.618    4    19   23   10    5    5   16  19       -23.0  
18419  0.615   19    28   47    9    5    1   17  39        -7.0  
18420  0.789   14    20   34    4    8    8   14  20       -12.0  

[18421 rows x 32 columns]

## Self Test - 3
- Find the number of observations and the number of variables in the dataset

In [66]:
nba_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18421 entries, 0 to 18420
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CITY               18421 non-null  object 
 1   TEAM_NAME          18421 non-null  object 
 2   TEAM_ID            18421 non-null  int64  
 3   NICKNAME           18421 non-null  object 
 4   STATE              18421 non-null  object 
 5   YEAR_FOUNDED       18421 non-null  int64  
 6   SEASON_ID          18421 non-null  int64  
 7   TEAM_ABBREVIATION  18421 non-null  object 
 8   GAME_ID            18421 non-null  int64  
 9   GAME_DATE          18421 non-null  object 
 10  MATCHUP            18421 non-null  object 
 11  WL                 18414 non-null  object 
 12  MIN                18421 non-null  int64  
 13  PTS                18421 non-null  int64  
 14  FGM                18421 non-null  int64  
 15  FGA                18421 non-null  int64  
 16  FG_PCT             184

The merged dataset is sorted by the criteria we use to merge the datasets. Thus, the NBA_Games dataset is currently sorted by "TEAM_ID." We may be interested to sort the data by other criteria, for example, the date of the game. 

#### We can do so by using the “sort_values” option. 

In our dataset, "GAME_ID" is created based on the date of the game. We can sort the games by “GAME_ID” and display the 20 most recent games.


In [74]:
nba_data.sort_values(by="GAME_ID", ascending=False).head(20)

CITY            TEAM_NAME     TEAM_ID   NICKNAME      STATE  \
13449  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15445         Utah            Utah Jazz  1610612762       Jazz       Utah   
1304     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
16076      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
15446         Utah            Utah Jazz  1610612762       Jazz       Utah   
1305     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
13450  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
16077      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
15447         Utah            Utah Jazz  1610612762       Jazz       Utah   
16078      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
1306     Cleveland  Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
13451  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15545         Utah            Utah Jazz  1610612762       Jazz       Utah   
99         Atlanta        Atlanta Hawks  1610612737      Hawks    Atlanta   
16173      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13551  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
15546         Utah            Utah Jazz  1610612762       Jazz       Utah   
16174      Memphis    Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13552  San Antonio    San Antonio Spurs  1610612759      Spurs      Texas   
100        Atlanta        Atlanta Hawks  1610612737      Hawks    Atlanta   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID GAME_DATE  ...  \
13449          1976      22019               SAS  1621900006  7/3/2019  ...   
15445          1974      22019               UTA  1621900006  7/3/2019  ...   
1304           1970      22019               CLE  1621900005  7/3/2019  ...   
16076          1995      22019               MEM  1621900005  7/3/2019  ...   
15446          1974      22019               UTA  1621900004  7/2/2019  ...   
1305           1970      22019               CLE  1621900004  7/2/2019  ...   
13450          1976      22019               SAS  1621900003  7/2/2019  ...   
16077          1995      22019               MEM  1621900003  7/2/2019  ...   
15447          1974      22019               UTA  1621900002  7/1/2019  ...   
16078          1995      22019               MEM  1621900002  7/1/2019  ...   
1306           1970      22019               CLE  1621900001  7/1/2019  ...   
13451          1976      22019               SAS  1621900001  7/1/2019  ...   
15545          1974      22018               UTA  1621800006  7/5/2018  ...   
99             1949      22018               ATL  1621800006  7/5/2018  ...   
16173          1995      22018               MEM  1621800005  7/5/2018  ...   
13551          1976      22018               SAS  1621800005  7/5/2018  ...   
15546          1974      22018               UTA  1621800004  7/3/2018  ...   
16174          1995      22018               MEM  1621800004  7/3/2018  ...   
13552          1976      22018               SAS  1621800003  7/3/2018  ...   
100            1949      22018               ATL  1621800003  7/3/2018  ...   

      FT_PCT OREB  DREB  REB  AST  STL  BLK  TOV  PF  PLUS_MINUS  
13449  0.615    9    29   38   12    5    3   11  14        -7.2  
15445  0.684   13    31   44   15    6    5   15  25         5.8  
1304   0.737    3    26   29   15    9    3   17  12       -13.0  
16076  0.692   11    36   47   19    8    5   14  19        13.0  
15446  0.750   11    34   45   18    3    4   14  18        14.6  
1305   0.682    6    27   33   13    6    0    6  14       -15.0  
13450  0.870   10    35   45   24   10    7   18  24        16.6  
16077  0.867   10    29   39   17    6    5   17  18       -15.8  
15447  0.462   14    27   41   16   10    7   19  26        -6.6  
16078  0.704    8    32   40   14   14    4   20  19         7.4  

## Missing Values

### Before we move on to doing any data analyses, we usually need to check if there is any missing value, that is, the source may have failed to collect some information. 

#### We can use the info() command which will return the total number of observations that have real values. By looking at these total numbers, we can see if there is any variable with missing value.

In [75]:
nba_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18421 entries, 0 to 18420
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CITY               18421 non-null  object 
 1   TEAM_NAME          18421 non-null  object 
 2   TEAM_ID            18421 non-null  int64  
 3   NICKNAME           18421 non-null  object 
 4   STATE              18421 non-null  object 
 5   YEAR_FOUNDED       18421 non-null  int64  
 6   SEASON_ID          18421 non-null  int64  
 7   TEAM_ABBREVIATION  18421 non-null  object 
 8   GAME_ID            18421 non-null  int64  
 9   GAME_DATE          18421 non-null  object 
 10  MATCHUP            18421 non-null  object 
 11  WL                 18414 non-null  object 
 12  MIN                18421 non-null  int64  
 13  PTS                18421 non-null  int64  
 14  FGM                18421 non-null  int64  
 15  FGA                18421 non-null  int64  
 16  FG_PCT             184

_The total number of rows is 18421, so there are missing values in variable WL, FG_PCT, FG3_PCT, and FT_PCT._

#### Detecting missing values
We can use the isnull() function and the notnull() function to detect where the missing values are.


In [82]:
nba_data.isnull()

CITY  TEAM_NAME  TEAM_ID  NICKNAME  STATE  YEAR_FOUNDED  SEASON_ID  \
0      False      False    False     False  False         False      False   
1      False      False    False     False  False         False      False   
2      False      False    False     False  False         False      False   
3      False      False    False     False  False         False      False   
4      False      False    False     False  False         False      False   
...      ...        ...      ...       ...    ...           ...        ...   
18416  False      False    False     False  False         False      False   
18417  False      False    False     False  False         False      False   
18418  False      False    False     False  False         False      False   
18419  False      False    False     False  False         False      False   
18420  False      False    False     False  False         False      False   

       TEAM_ABBREVIATION  GAME_ID  GAME_DATE  ...  FT_PCT   OREB   DREB  \
0                  False    False      False  ...   False  False  False   
1                  False    False      False  ...   False  False  False   
2                  False    False      False  ...   False  False  False   
3                  False    False      False  ...   False  False  False   
4                  False    False      False  ...   False  False  False   
...                  ...      ...        ...  ...     ...    ...    ...   
18416              False    False      False  ...   False  False  False   
18417              False    False      False  ...   False  False  False   
18418              False    False      False  ...   False  False  False   
18419              False    False      False  ...   False  False  False   
18420              False    False      False  ...   False  False  False   

         REB    AST    STL    BLK    TOV     PF  PLUS_MINUS  
0      False  False  False  False  False  False       False  
1      False  False  False  False  False  False       False  
2      False  False  False  False  False  False       False  
3      False  False  False  False  False  False       False  
4      False  False  False  False  False  False       False  
...      ...    ...    ...    ...    ...    ...         ...  
18416  False  False  False  False  False  False       False  
18417  False  False  False  False  False  False       False  
18418  False  False  False  False  False  False       False  
18419  False  False  False  False  False  False       False  
18420  False  False  False  False  False  False       False  

[18421 rows x 32 columns]

### Handling Missing Values

There are two main approaches to handle missing values. 
- First, we can simply drop the observations with missing value.

#### Drop observations with missing value in the variable "FG_PCT"

In [89]:
display(nba_data[nba_data["FG_PCT"].isnull()])

nba_data = nba_data[nba_data["FG_PCT"].notnull()]
nba_data

CITY           TEAM_NAME     TEAM_ID NICKNAME  \
2736      Chicago       Chicago Bulls  1610612741    Bulls   
16713  Washington  Washington Wizards  1610612764  Wizards   

                      STATE  YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION  \
2736               Illinois          1966      12017               CHI   
16713  District of Columbia          1961      22019               WAS   

          GAME_ID  GAME_DATE  ... FT_PCT OREB  DREB  REB  AST  STL  BLK  TOV  \
2736     11700012  10/3/2017  ...    NaN    0     0    0    0    0    0    0   
16713  1521900060  7/11/2019  ...    NaN    0     0    0    0    0    0    0   

       PF  PLUS_MINUS  
2736    0         0.0  
16713   0        -0.8  

[2 rows x 32 columns]

CITY          TEAM_NAME     TEAM_ID NICKNAME           STATE  \
0        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
1        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
2        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
3        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
4        Atlanta      Atlanta Hawks  1610612737    Hawks         Atlanta   
...          ...                ...         ...      ...             ...   
18416  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18417  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18418  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18419  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   
18420  Charlotte  Charlotte Hornets  1610612766  Hornets  North Carolina   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ...  \
0              1949      22019               ATL  1521900072  7/12/2019  ...   
1              1949      22019               ATL  1521900060  7/11/2019  ...   
2              1949      22019               ATL  1521900042   7/9/2019  ...   
3              1949      22019               ATL  1521900023   7/7/2019  ...   
4              1949      22019               ATL  1521900023   7/7/2019  ...   
...             ...        ...               ...         ...        ...  ...   
18416          1988      22014               CHA  1521400046  7/17/2014  ...   
18417          1988      22014               CHA  1521400038  7/16/2014  ...   
18418          1988      22014               CHA  1521400023  7/14/2014  ...   
18419          1988      22014               CHA  1521400017  7/13/2014  ...   
18420          1988      22014               CHA  1521400004  7/11/2014  ...   

      FT_PCT OREB  DREB  REB  AST  STL  BLK  TOV  PF  PLUS_MINUS  
0      0.850   13    23   36   14   15    3   12  24         8.0  
1      0.700    9    28   37   19   10    8   22  25        -5.0  
2      0.708    7    27   34   17    5    5   18  21        18.2  
3      0.500    1     2    3    1    0    1    4   2         0.0  
4      0.625    9    27   36    7    7   10   18  28       -24.0  
...      ...  ...   ...  ...  ...  ...  ...  ...  ..         ...  
18416  0.762   12    26   38   19    7    2   11  22        29.0  
18417  0.667    7    33   40   10    7    2   21  31         1.0  
18418  0.618    4    19   23   10    5    5   16  19       -23.0  
18419  0.615   19    28   47    9    5    1   17  39        -7.0  
18420  0.789   14    20   34    4    8    8   14  20       -12.0  

[18419 rows x 32 columns]

- Second, we can replace the missing values with valid values (Imputation), such as mean and median.

#### We can use the fillna() command to replace missing values with the mean or the median of the variable.

In [92]:
nba_data = nba_data.fillna(nba_data.mean())
nba_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18419 entries, 0 to 18420
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CITY               18419 non-null  object 
 1   TEAM_NAME          18419 non-null  object 
 2   TEAM_ID            18419 non-null  int64  
 3   NICKNAME           18419 non-null  object 
 4   STATE              18419 non-null  object 
 5   YEAR_FOUNDED       18419 non-null  int64  
 6   SEASON_ID          18419 non-null  int64  
 7   TEAM_ABBREVIATION  18419 non-null  object 
 8   GAME_ID            18419 non-null  int64  
 9   GAME_DATE          18419 non-null  object 
 10  MATCHUP            18419 non-null  object 
 11  WL                 18414 non-null  object 
 12  MIN                18419 non-null  int64  
 13  PTS                18419 non-null  int64  
 14  FGM                18419 non-null  int64  
 15  FGA                18419 non-null  int64  
 16  FG_PCT             184

/tmp/ipykernel_406349/3628645146.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  nba_data = nba_data.fillna(nba_data.mean())


## Creating variables
We can create a variable equals to the total number of goals made.


In [95]:
nba_data['GM'] = nba_data['FGM'] + nba_data['FG3M'] + nba_data['FTM']
nba_data[["FGM", "FG3M", "FTM", "GM"]]

FGM  FG3M  FTM  GM
0       27     9   17  53
1       26    12    7  45
2       31     8   17  56
3        2     1    1   4
4       18     4   20  42
...    ...   ...  ...  ..
18416   38    12   16  66
18417   25     6   26  57
18418   22     7   21  50
18419   20     9   16  45
18420   20     3   15  38

[18419 rows x 4 columns]

## Self Test - 4
- Create a variable called "GA" equals to the total number of goals attempted.

In [99]:
nba_data["GA"] = nba_data["FGA"] + nba_data["FG3A"] + nba_data["FTA"]
nba_data[["FGA", "FG3A", "FTA", "GA"]]

FGA  FG3A  FTA   GA
0       79    32   20  131
1       68    29   10  107
2       60    21   24  105
3        8     3    2   13
4       62    22   32  116
...    ...   ...  ...  ...
18416   76    20   21  117
18417   62    16   39  117
18418   54    18   34  106
18419   70    30   26  126
18420   71    28   19  118

[18419 rows x 4 columns]

### Create variables based on conditions
- We can create a variable conditional on the value of another variable.

For example, we can create a variable "RESULT" that equals to 1 if the team won the game and 0 otherwise. The result of the game can be captured in the points of the team receive, whether it was positive or negative.

In [100]:
nba_data.columns

Index(['CITY', 'TEAM_NAME', 'TEAM_ID', 'NICKNAME', 'STATE', 'YEAR_FOUNDED',
       'SEASON_ID', 'TEAM_ABBREVIATION', 'GAME_ID', 'GAME_DATE', 'MATCHUP',
       'WL', 'MIN', 'PTS', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT',
       'FTM', 'FTA', 'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK',
       'TOV', 'PF', 'PLUS_MINUS', 'GM', 'GA'],
      dtype='object')

In [103]:
nba_data["RESULT"] = np.where(nba_data["PLUS_MINUS"] > 0, "W", "L")
nba_data.head()

CITY      TEAM_NAME     TEAM_ID NICKNAME    STATE  YEAR_FOUNDED  \
0  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
1  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
2  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
3  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
4  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   

   SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ... REB AST  STL  BLK  \
0      22019               ATL  1521900072  7/12/2019  ...  36  14   15    3   
1      22019               ATL  1521900060  7/11/2019  ...  37  19   10    8   
2      22019               ATL  1521900042   7/9/2019  ...  34  17    5    5   
3      22019               ATL  1521900023   7/7/2019  ...   3   1    0    1   
4      22019               ATL  1521900023   7/7/2019  ...  36   7    7   10   

   TOV  PF  PLUS_MINUS  GM   GA  RESULT  
0   12  24         8.0  53  131       W  
1   22  25        -5.0  45  107       L  
2   18  21        18.2  56  105       W  
3    4   2         0.0   4   13       L  
4   18  28       -24.0  42  116       L  

[5 rows x 35 columns]

In [104]:
nba_data.drop(["RESULT"], axis=1, inplace=True)
nba_data.head()

CITY      TEAM_NAME     TEAM_ID NICKNAME    STATE  YEAR_FOUNDED  \
0  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
1  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
2  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
3  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
4  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   

   SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ... DREB REB  AST  STL  \
0      22019               ATL  1521900072  7/12/2019  ...   23  36   14   15   
1      22019               ATL  1521900060  7/11/2019  ...   28  37   19   10   
2      22019               ATL  1521900042   7/9/2019  ...   27  34   17    5   
3      22019               ATL  1521900023   7/7/2019  ...    2   3    1    0   
4      22019               ATL  1521900023   7/7/2019  ...   27  36    7    7   

   BLK  TOV  PF  PLUS_MINUS  GM   GA  
0    3   12  24         8.0  53  131  
1    8   22  25        -5.0  45  107  
2    5   18  21        18.2  56  105  
3    1    4   2         0.0   4   13  
4   10   18  28       -24.0  42  116  

[5 rows x 34 columns]

### Create a variable within group
In the dataset, each game has two observations, one represents the statistics of the home team, one represents those of the away team. Both observations have the same GAME_ID. We can create a variable "POINT_DIFF" that equals the difference between the points earned by the two teams.

We will first sort the data not only by the "GAME_ID" but also by the result "WL".


In [125]:
nba_data.sort_values(by=["GAME_ID", "WL"], inplace=True)
nba_data["POINT_DIFF"] = nba_data.groupby(["GAME_ID"])["PTS"].diff()
nba_data

CITY              TEAM_NAME     TEAM_ID   NICKNAME      STATE  \
14741  Oklahoma City  Oklahoma City Thunder  1610612760    Thunder   Oklahoma   
10949        Indiana         Indiana Pacers  1610612754     Pacers    Indiana   
3142         Chicago          Chicago Bulls  1610612741      Bulls   Illinois   
5762         Houston        Houston Rockets  1610612745    Rockets      Texas   
2546     New Orleans   New Orleans Pelicans  1610612740   Pelicans  Louisiana   
...              ...                    ...         ...        ...        ...   
15446           Utah              Utah Jazz  1610612762       Jazz       Utah   
1304       Cleveland    Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
16076        Memphis      Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13449    San Antonio      San Antonio Spurs  1610612759      Spurs      Texas   
15445           Utah              Utah Jazz  1610612762       Jazz       Utah   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ...  \
14741          1967      12013               OKC    11300001  10/5/2013  ...   
10949          1976      12013               IND    11300002  10/5/2013  ...   
3142           1966      12013               CHI    11300002  10/5/2013  ...   
5762           1967      12013               HOU    11300003  10/5/2013  ...   
2546           2002      12013               NOP    11300003  10/5/2013  ...   
...             ...        ...               ...         ...        ...  ...   
15446          1974      22019               UTA  1621900004   7/2/2019  ...   
1304           1970      22019               CLE  1621900005   7/3/2019  ...   
16076          1995      22019               MEM  1621900005   7/3/2019  ...   
13449          1976      22019               SAS  1621900006   7/3/2019  ...   
15445          1974      22019               UTA  1621900006   7/3/2019  ...   

      REB AST  STL  BLK  TOV  PF  PLUS_MINUS  GM   GA  POINT_DIFF  
14741  52  22    9    8   20  26        13.0  59  116         NaN  
10949  38  15   12    8   15  23        -6.0  49  122         NaN  
3142   56  20    5   10   23  25         6.0  54  114         6.0  
5762   35  24    9    4   22  27        -1.0  76  128         NaN  
2546   33  17   12    4   15  32         1.0  74  136         1.0  
...    ..  ..  ...  ...  ...  ..         ...  ..  ...         ...  
15446  45  18    3    4   14  18        14.6  52  111        15.0  
1304   29  15    9    3   17  12       -13.0  46  114         NaN  
16076  47  19    8    5   14  19        13.0  50  115        13.0  
13449  38  12    5    3   11  14        -7.2  52  113         NaN  
15445  44  15    6    5   15  25         5.8  54  124         3.0  

[18419 rows x 35 columns]

The "POINT_DIFF" variable only has the point difference for the winning team, we need to impute the point difference for the losing team as well.


In [127]:
nba_data["POINT_DIFF"] = nba_data['POINT_DIFF'].fillna(nba_data.groupby('GAME_ID')['POINT_DIFF'].transform('mean'))
nba_data

CITY              TEAM_NAME     TEAM_ID   NICKNAME      STATE  \
14741  Oklahoma City  Oklahoma City Thunder  1610612760    Thunder   Oklahoma   
10949        Indiana         Indiana Pacers  1610612754     Pacers    Indiana   
3142         Chicago          Chicago Bulls  1610612741      Bulls   Illinois   
5762         Houston        Houston Rockets  1610612745    Rockets      Texas   
2546     New Orleans   New Orleans Pelicans  1610612740   Pelicans  Louisiana   
...              ...                    ...         ...        ...        ...   
15446           Utah              Utah Jazz  1610612762       Jazz       Utah   
1304       Cleveland    Cleveland Cavaliers  1610612739  Cavaliers       Ohio   
16076        Memphis      Memphis Grizzlies  1610612763  Grizzlies  Tennessee   
13449    San Antonio      San Antonio Spurs  1610612759      Spurs      Texas   
15445           Utah              Utah Jazz  1610612762       Jazz       Utah   

       YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  GAME_DATE  ...  \
14741          1967      12013               OKC    11300001  10/5/2013  ...   
10949          1976      12013               IND    11300002  10/5/2013  ...   
3142           1966      12013               CHI    11300002  10/5/2013  ...   
5762           1967      12013               HOU    11300003  10/5/2013  ...   
2546           2002      12013               NOP    11300003  10/5/2013  ...   
...             ...        ...               ...         ...        ...  ...   
15446          1974      22019               UTA  1621900004   7/2/2019  ...   
1304           1970      22019               CLE  1621900005   7/3/2019  ...   
16076          1995      22019               MEM  1621900005   7/3/2019  ...   
13449          1976      22019               SAS  1621900006   7/3/2019  ...   
15445          1974      22019               UTA  1621900006   7/3/2019  ...   

      REB AST  STL  BLK  TOV  PF  PLUS_MINUS  GM   GA  POINT_DIFF  
14741  52  22    9    8   20  26        13.0  59  116         NaN  
10949  38  15   12    8   15  23        -6.0  49  122         6.0  
3142   56  20    5   10   23  25         6.0  54  114         6.0  
5762   35  24    9    4   22  27        -1.0  76  128         1.0  
2546   33  17   12    4   15  32         1.0  74  136         1.0  
...    ..  ..  ...  ...  ...  ..         ...  ..  ...         ...  
15446  45  18    3    4   14  18        14.6  52  111        15.0  
1304   29  15    9    3   17  12       -13.0  46  114        13.0  
16076  47  19    8    5   14  19        13.0  50  115        13.0  
13449  38  12    5    3   11  14        -7.2  52  113         3.0  
15445  44  15    6    5   15  25         5.8  54  124         3.0  

[18419 rows x 35 columns]

In [156]:
nba_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18419 entries, 14741 to 15445
Data columns (total 35 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CITY               18419 non-null  object 
 1   TEAM_NAME          18419 non-null  object 
 2   TEAM_ID            18419 non-null  int64  
 3   NICKNAME           18419 non-null  object 
 4   STATE              18419 non-null  object 
 5   YEAR_FOUNDED       18419 non-null  int64  
 6   SEASON_ID          18419 non-null  int64  
 7   TEAM_ABBREVIATION  18419 non-null  object 
 8   GAME_ID            18419 non-null  int64  
 9   GAME_DATE          18419 non-null  object 
 10  MATCHUP            18419 non-null  object 
 11  WL                 18414 non-null  object 
 12  MIN                18419 non-null  int64  
 13  PTS                18419 non-null  int64  
 14  FGM                18419 non-null  int64  
 15  FGA                18419 non-null  int64  
 16  FG_PCT            

In [189]:
ids = nba_data["GAME_ID"].value_counts() < 2
ids_df = ids.reset_index().rename(columns={"index": "id", "GAME_ID": "flag"})
nba_data[nba_data["GAME_ID"].isin(ids_df[ids_df["flag"]==True]["id"])]

CITY               TEAM_NAME     TEAM_ID      NICKNAME  \
14741  Oklahoma City   Oklahoma City Thunder  1610612760       Thunder   
11591   Philadelphia      Philadelphia 76ers  1610612755         76ers   
8521       Minnesota  Minnesota Timberwolves  1610612750  Timberwolves   
12178        Phoenix            Phoenix Suns  1610612756          Suns   
606          Atlanta           Atlanta Hawks  1610612737         Hawks   
...              ...                     ...         ...           ...   
7332       Milwaukee         Milwaukee Bucks  1610612749         Bucks   
12853     Sacramento        Sacramento Kings  1610612758         Kings   
14133  Oklahoma City   Oklahoma City Thunder  1610612760       Thunder   
11618        Phoenix            Phoenix Suns  1610612756          Suns   
10351        Indiana          Indiana Pacers  1610612754        Pacers   

              STATE  YEAR_FOUNDED  SEASON_ID TEAM_ABBREVIATION     GAME_ID  \
14741      Oklahoma          1967      12013               OKC    11300001   
11591  Pennsylvania          1949      12013               PHI    11300005   
8521      Minnesota          1989      12013               MIN    11300008   
12178       Arizona          1968      12013               PHX    11300009   
606         Atlanta          1949      12013               ATL    11300019   
...             ...           ...        ...               ...         ...   
7332      Wisconsin          1968      22019               MIL  1521900057   
12853    California          1948      22019               SAC  1521900063   
14133      Oklahoma          1967      22019               OKC  1521900067   
11618       Arizona          1968      22019               PHX  1521900068   
10351       Indiana          1976      22019               IND  1521900070   

       GAME_DATE  ... REB AST  STL  BLK  TOV  PF  PLUS_MINUS  GM   GA  \
14741  10/5/2013  ...  52  22    9    8   20  26        13.0  59  116   
11591  10/6/2013  ...  39  17   23    2   26  34         2.0  74  138   
8521   10/7/2013  ...  46  18    6    6   11  22        -2.0  66  154   
12178  10/7/2013  ...  47  30   16    9   16  21        41.0  75  127   
606    10/8/2013  ...  44  24    8   12   17  28         2.0  56  120   
...          ...  ...  ..  ..  ...  ...  ...  ..         ...  ..  ...   
7332   7/10/2019  ...  29  15    8    0   12  20        17.0  58  123   
12853  7/11/2019  ...  45  12    8   10   20  22        -0.6  48  115   
14133  7/12/2019  ...  25  18    7    8   13  11         3.4  41   99   
11618  7/12/2019  ...  30  18   12    1   12  28        29.0  58   98   
10351  7/12/2019  ...  39  12    8    1    9  15         5.0  53  119   

       POINT_DIFF  
14741         NaN  
11591         NaN  
8521          NaN  
12178         NaN  
606           NaN  
...           ...  
7332          NaN  
12853         NaN  
14133         NaN  
11618         NaN  
10351         NaN  

[635 rows x 35 columns]

- We can also drop all observations with missing value in at least one variable using the "dropna()" command.


In [196]:
nba_data=nba_data.dropna()
nba_data.shape

(17779, 35)

### Creating new dataframe

#### Create a new dataframe that aggregates information by group

Sometimes we may want to work with season level data rather than team level data. We can create a new dataset that includes aggregate information of team statistics in each season.

In [212]:
nba_team_stats = nba_data.groupby(["TEAM_NAME", "SEASON_ID"])[['PTS','FGM','FGA','FG_PCT','FG3M','FG3A','FG3_PCT','FTM','FTA','FT_PCT','OREB','DREB','REB','AST','STL','BLK','TOV','PF','PLUS_MINUS']].sum()
nba_team_stats.head(20)

PTS   FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT   FTM  \
TEAM_NAME      SEASON_ID                                                        
Atlanta Hawks  12013       523   193   459   2.520    34   131    1.576   103   
               12014       708   247   554   3.122    66   182    2.549   148   
               12015       652   226   548   2.896    59   176    2.340   141   
               12016       686   261   593   3.083    50   159    2.140   114   
               12017       480   167   410   2.045    51   156    1.611    95   
               12018       563   206   445   2.312    63   189    1.659    88   
               22012      2074   795  1701   9.824   155   454    7.206   329   
               22013      8272  3061  6719  37.846   747  2085   29.419  1403   
               22014      8786  3253  7017  40.399   848  2256   32.719  1432   
               22015      8701  3265  7163  39.250   833  2403   29.818  1338   
               22016      8736  3219  7142  38.784   760  2214   29.491  1538   
               22017      8683  3206  7208  37.909   919  2567   30.388  1352   
               22018      9742  3555  7987  39.210  1121  3226   30.483  1511   
               22019       381   127   342   1.873    43   136    1.571    84   
               42012       536   196   464   2.533    37   112    1.900   107   
               42013       647   212   551   2.729    79   230    2.483   144   
               42014      1568   590  1369   6.907   145   458    5.030   243   
               42015      1001   366   848   4.337   106   318    3.332   163   
               42016       633   225   507   2.662    50   164    1.798   133   
Boston Celtics 12013       754   276   654   3.404    54   186    2.313   148   

                           FTA  FT_PCT  OREB  DREB   REB   AST  STL  BLK  \
TEAM_NAME      SEASON_ID                                                   
Atlanta Hawks  12013       140   4.477    44   184   228   130   61   22   
               12014       197   5.262    63   237   300   178   48   35   
               12015       175   5.630    53   271   324   141   57   41   
               12016       153   5.253    71   263   334   184   52   40   
               12017       125   3.798    40   169   209   113   40   15   
               12018       124   3.641    49   175   224   120   49   31   
               22012       444  15.640   191   681   872   506  178   68   
               22013      1799  64.601   731  2532  3263  2005  686  327   
               22014      1840  67.699   754  2728  3482  2171  786  396   
               22015      1716  66.851   715  2868  3583  2143  789  517   
               22016      2115  62.499   883  2902  3785  1985  721  424   
               22017      1737  66.456   775  2768  3543  1971  664  355   
               22018      2034  65.306  1015  3010  4025  2193  735  439   
               22019       112   3.768    47   135   182    70   48   32   
               42012       157   3.990    50   180   230   115   44   23   
               42013       174   5.837    64   211   275   118   46   32   
               42014       311  12.387   157   542   699   381  136   79   
               42015       205   7.895    89   354   443   226   75   76   
               42016       174   4.552    68   198   266   134   36   20   
Boston Celtics 12013       188   6.358    93   227   320   168   77   38   

                           TOV    PF  PLUS_MINUS  
TEAM_NAME      SEASON_ID                          
Atlanta Hawks  12013       118   112       -75.0  
               12014       116   170        15.0  
               12015       122   143        -5.0  
               12016       112   154        41.0  
               12017        95    91       -17.6  
               12018       104   157       -18.0  
               22012       293   367       -15.0  
               22013      1220  1610       -81.0  
               22014      1183  1577       433.0  
          

Notice that the newly created dataset has two levels of index, the "TEAM_ID" and "SEASON_ID"

#### If we want to convert these two indexes back as variables, we can use the "reset_index" command.


In [208]:
nba_team_stats.reset_index(inplace=True)
nba_team_stats

index           TEAM_NAME  SEASON_ID   PTS  FGM   FGA  FG_PCT  FG3M  \
0        0       Atlanta Hawks      12013   523  193   459   2.520    34   
1        1       Atlanta Hawks      12014   708  247   554   3.122    66   
2        2       Atlanta Hawks      12015   652  226   548   2.896    59   
3        3       Atlanta Hawks      12016   686  261   593   3.083    50   
4        4       Atlanta Hawks      12017   480  167   410   2.045    51   
..     ...                 ...        ...   ...  ...   ...     ...   ...   
510    510  Washington Wizards      22019   389  141   371   1.897    29   
511    511  Washington Wizards      42013   994  379   877   4.747    62   
512    512  Washington Wizards      42014  1011  373   845   4.461    94   
513    513  Washington Wizards      42016  1405  521  1139   5.959   109   
514    514  Washington Wizards      42017   643  238   511   2.793    48   

     FG3A  FG3_PCT  ...  FT_PCT  OREB  DREB  REB  AST  STL  BLK  TOV   PF  \
0     131    1.576  ...   4.477    44   184  228  130   61   22  118  112   
1     182    2.549  ...   5.262    63   237  300  178   48   35  116  170   
2     176    2.340  ...   5.630    53   271  324  141   57   41  122  143   
3     159    2.140  ...   5.253    71   263  334  184   52   40  112  154   
4     156    1.611  ...   3.798    40   169  209  113   40   15   95   91   
..    ...      ...  ...     ...   ...   ...  ...  ...  ...  ...  ...  ...   
510   124    1.183  ...   3.755    50   141  191   74   32   21   61  102   
511   185    3.654  ...   7.541   129   365  494  215   82   62  135  245   
512   233    3.934  ...   7.113   116   356  472  231   71   53  135  200   
513   344    4.170  ...  10.329   145   414  559  296  103   86  170  299   
514   131    2.239  ...   5.175    51   189  240  140   47   31   74  126   

     PLUS_MINUS  
0         -75.0  
1          15.0  
2          -5.0  
3          41.0  
4         -17.6  
..          ...  
510       -12.4  
511        10.0  
512        39.0  
513         0.0  
514       -12.0  

[515 rows x 22 columns]

### We can create a variable that equals to the total number of observations within a specified group using the size() command.
- Create a variable that equals to the total number of games played by a team in each season, name this variable "GAME_COUNT".

In [213]:
game_count = nba_data.groupby(["TEAM_NAME", "SEASON_ID"]).size().reset_index(name="GAME_COUNT")
game_count

TEAM_NAME  SEASON_ID  GAME_COUNT
0         Atlanta Hawks      12013           6
1         Atlanta Hawks      12014           7
2         Atlanta Hawks      12015           7
3         Atlanta Hawks      12016           7
4         Atlanta Hawks      12017           5
..                  ...        ...         ...
510  Washington Wizards      22019           5
511  Washington Wizards      42013          11
512  Washington Wizards      42014          10
513  Washington Wizards      42016          13
514  Washington Wizards      42017           6

[515 rows x 3 columns]

In [217]:
nba_data.groupby(["TEAM_NAME", "SEASON_ID"]).size()

TEAM_NAME           SEASON_ID
Atlanta Hawks       12013         6
                    12014         7
                    12015         7
                    12016         7
                    12017         5
                                 ..
Washington Wizards  22019         5
                    42013        11
                    42014        10
                    42016        13
                    42017         6
Length: 515, dtype: int64

In [220]:
nba_data[(nba_data["TEAM_NAME"] == "Atlanta Hawks") & (nba_data["SEASON_ID"] == 12013)]

CITY      TEAM_NAME     TEAM_ID NICKNAME    STATE  YEAR_FOUNDED  \
607  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
605  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
604  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
603  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
602  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   
601  Atlanta  Atlanta Hawks  1610612737    Hawks  Atlanta          1949   

     SEASON_ID TEAM_ABBREVIATION   GAME_ID   GAME_DATE  ... REB AST  STL  BLK  \
607      12013               ATL  11300012   10/7/2013  ...  41  20   11    6   
605      12013               ATL  11300043  10/13/2013  ...  28  19   11    5   
604      12013               ATL  11300068  10/17/2013  ...  42  30   12    1   
603      12013               ATL  11300079  10/20/2013  ...  43  20   12    6   
602      12013               ATL  11300089  10/22/2013  ...  32  27    4    4   
601      12013               ATL  11300101  10/23/2013  ...  42  14   11    0   

     TOV  PF  PLUS_MINUS  GM   GA  POINT_DIFF  
607   21  19        -5.0  55  123         5.0  
605   24  18       -32.0  47  109        32.0  
604   16  18        -2.0  65  136         2.0  
603   20  21        -8.0  54  130         8.0  
602   19  19       -18.0  53  109        18.0  
601   18  17       -10.0  56  123        10.0  

[6 rows x 35 columns]